sender_test.go

```diff
@@ -2,32 +2,769 @@ package amqp
+
+func TestSenderSendRejected(t *testing.T) {
+	responder := func(req frames.FrameBody) ([]byte, error) {
+		switch tt := req.(type) {
+		case *mocks.AMQPProto:
+			return []byte{'A', 'M', 'Q', 'P', 0, 1, 0, 0}, nil
+		case *frames.PerformOpen:
+			return mocks.PerformOpen("container")
+		case *frames.PerformBegin:
+			return mocks.PerformBegin(0)
+		case *frames.PerformEnd:
+			return mocks.PerformEnd(0, nil)
+		case *frames.PerformAttach:
+			return mocks.SenderAttach(0, tt.Name, 0, encoding.ModeUnsettled)
+		case *frames.PerformTransfer:
+			return mocks.PerformDisposition(encoding.RoleReceiver, 0, *tt.DeliveryID, &encoding.StateRejected{
+				Error: &Error{
+					Condition:   "rejected",
+					Description: "didn't like it",
+				},
+			})
+		case *frames.PerformDetach:
+			return mocks.PerformDetach(0, 0, nil)
+		default:
+			return nil, fmt.Errorf("unhandled frame %T", req)
+		}
+	}
+	netConn := mocks.NewNetConn(responder)
+
+	client, err := New(netConn)
+	require.NoError(t, err)
+
+	session, err := client.NewSession()
+	require.NoError(t, err)
+	time.Sleep(100 * time.Millisecond)
+	snd, err := session.NewSender()
+	require.NoError(t, err)
+
+	sendInitialFlowFrame(t, netConn, 0, 100)
+	time.Sleep(100 * time.Millisecond)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+	err = snd.Send(ctx, NewMessage([]byte("test")))
+	cancel()
+	var asErr *Error
+	if !errors.As(err, &asErr) {
+		t.Fatalf("unexpected error type %T", err)
+	}
+	require.Equal(t, encoding.ErrorCondition("rejected"), asErr.Condition)
+
+	time.Sleep(100 * time.Millisecond)
+	require.NoError(t, client.Close())
+}
+
+func TestSenderSendDetached(t *testing.T) {
+	responder := func(req frames.FrameBody) ([]byte, error) {
+		switch tt := req.(type) {
+		case *mocks.AMQPProto:
+			return []byte{'A', 'M', 'Q', 'P', 0, 1, 0, 0}, nil
+		case *frames.PerformOpen:
+			return mocks.PerformOpen("container")
+		case *frames.PerformBegin:
+			return mocks.PerformBegin(0)
+		case *frames.PerformEnd:
+			return mocks.PerformEnd(0, nil)
+		case *frames.PerformAttach:
+			return mocks.SenderAttach(0, tt.Name, 0, encoding.ModeUnsettled)
+		case *frames.PerformTransfer:
+			return mocks.PerformDetach(0, 0, &Error{
+				Condition:   "detached",
+				Description: "server exploded",
+			})
+		case *frames.PerformDetach:
+			return mocks.PerformDetach(0, 0, nil)
+		default:
+			return nil, fmt.Errorf("unhandled frame %T", req)
+		}
+	}
+	netConn := mocks.NewNetConn(responder)
+
+	client, err := New(netConn)
+	require.NoError(t, err)
+
+	session, err := client.NewSession()
+	require.NoError(t, err)
+	time.Sleep(100 * time.Millisecond)
+	snd, err := session.NewSender()
+	require.NoError(t, err)
+
+	sendInitialFlowFrame(t, netConn, 0, 100)
+	time.Sleep(100 * time.Millisecond)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+	err = snd.Send(ctx, NewMessage([]byte("test")))
+	cancel()
+	var asErr *DetachError
+	if !errors.As(err, &asErr) {
```

Comment:
Not for this PR: do you think we'll start doing some sort of NonRetriable()  method at some point? It'd be nice to avoid having to check both sentinel errors _and_ doing .As(), since they're all originating from the same library. (ie, it'd be nice to just have one style)

Reply:
Opened https://github.com/Azure/go-amqp/issues/98 to track this.